File: Arrays/Hard/4/4.c
#include <limits.h>

double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size) {
    if (nums1Size > nums2Size) {
        return findMedianSortedArrays(nums2, nums2Size, nums1, nums1Size);
    }

    int total = nums1Size + nums2Size;
    int half = (total + 1) / 2;

    int left = 0;
    int right = nums1Size;

    while (left <= right) {
        int mid1 = (left + right) / 2;
        int mid2 = half - mid1;

        int maxLeft1 = (mid1 == 0) ? INT_MIN : nums1[mid1 - 1];
        int minRight1 = (mid1 == nums1Size) ? INT_MAX : nums1[mid1];
        int maxLeft2 = (mid2 == 0) ? INT_MIN : nums2[mid2 - 1];
        int minRight2 = (mid2 == nums2Size) ? INT_MAX : nums2[mid2];

        if (maxLeft1 <= minRight2 && maxLeft2 <= minRight1) {
            if (total % 2 == 0) {
                return (double)(fmax(maxLeft1, maxLeft2) + fmin(minRight1, minRight2)) / 2.0;
            } else {
                return (double)fmax(maxLeft1, maxLeft2);
            }
        } else if (maxLeft1 > minRight2) {
            right = mid1 - 1;
        } else {
            left = mid1 + 1;
        }
    }
    return 0.0;
}

File: Arrays/Hard/4/4.py
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        a, b = nums1, nums2
        total = len(nums1) + len(nums2)
        partition = total // 2

        if len(b) < len(a):
            a, b = b, a

        l, r = 0, len(a) - 1
        while True:
            midA = (l + r) // 2
            midB = partition - midA - 2

            aL = a[midA] if midA >= 0 else float("-infinity")
            aR = a[midA + 1] if (midA + 1) < len(a) else float("-infinity") 
            bL = b[midB] if midB >= 0 else float("-infinity")
            bR = b[midB + 1] if (midB + 1) < len(a) else float("-infinity")

            if aL <= bR and bL <= aR:
                if total % 2:
                    return min(aR, bR)
                return (max(aL, bL) + min(aR, bR)) / 2
            elif aL > bR:
                r = midA - 1
            else:
                l = midA + 1